src/raftstore/store/snap.rs

```diff
@@ -0,0 +1,284 @@
+        let len = try!(reader.metadata()).len();
+        if len < 4 {
+            return Err(io::Error::new(ErrorKind::InvalidInput, format!("file length {} < 4", len)));
+        }
+        let to_read = len as usize - 4;
+        let mut total_read = 0;
+        let mut buffer = vec![0; 4098];
+        loop {
+            let read = try!(reader.read(&mut buffer));
+            if total_read + read >= to_read {
+                digest.write(&buffer[..to_read - total_read]);
+                try!(reader.seek(SeekFrom::End(-4)));
+                break;
+            }
+            digest.write(&buffer);
+            total_read += read;
+        }
+        let sum = try!(reader.read_u32::<BigEndian>());
+        if sum != digest.sum32() {
+            return Err(io::Error::new(ErrorKind::InvalidData,
+                                      format!("crc not correct: {} != {}", sum, digest.sum32())));
+        }
+        Ok(())
+    }
+
+    pub fn exists(&self) -> bool {
+        self.file.exists() && self.file.is_file()
+    }
+
+    pub fn delete(&self) {
+        if let Err(e) = self.try_delete() {
+            error!("failed to delete {}: {:?}", self.path().display(), e);
+        }
+    }
+
+    pub fn try_delete(&self) -> io::Result<()> {
+        fs::remove_file(self.path())
+    }
+
+    /// Use the content in temporary files replace the target file.
+    ///
+    /// Please note that this method can only be called once.
+    pub fn save(&mut self) -> io::Result<()> {
+        if let Some((mut f, path)) = self.tmp_file.take() {
+            try!(f.write_u32::<BigEndian>(self.digest.sum32()));
+            try!(f.flush());
+            try!(fs::rename(path, self.file.as_path()));
+        }
+        Ok(())
+    }
+
+    pub fn path(&self) -> &Path {
+        self.file.as_path()
+    }
+}
+
+impl Write for SnapFile {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        if self.tmp_file.is_none() {
+            return Ok(0);
+        }
+        let written = try!(self.tmp_file.as_mut().unwrap().0.write(buf));
+        self.digest.write(&buf[..written]);
+        Ok(written)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        if self.tmp_file.is_none() {
+            return Ok(());
+        }
+        self.tmp_file.as_mut().unwrap().0.flush()
+    }
+}
+
+impl Drop for SnapFile {
+    fn drop(&mut self) {
+        if let Some((_, path)) = self.tmp_file.take() {
+            if let Err(e) = fs::remove_file(&path) {
+                warn!("failed to delete temporary file {}: {:?}", path, e);
+            }
+        }
+    }
+}
+
+/// `SnapManagerCore` trace all current processing snapshots.
+pub struct SnapManagerCore {
+    // directory to store snapfile.
+    base: String,
+    // bool indicate if the snapshot is being generated by the peer on current store.
+    // usize indicate the times the snapkey has been registered.
+    registry: HashMap<SnapKey, (bool, usize)>,
```

Comment:
true means sending? false means not sending?
Do we need enum to cover applying later?